core/src/service/file_metadata_service.rs

```diff
@@ -1,140 +1,392 @@
+    fn sync(db: &Db) -> Result<Vec<FileMetadata>, Error>;
+    fn create(db: &Db, name: String, path: String) -> Result<FileMetadata, Error>;
 }
 
 pub struct FileMetadataServiceImpl<
     FileMetadataDb: FileMetadataRepo,
+    FileDb: FileRepo,
     AccountDb: AccountRepo,
-    C: Client,
+    ApiClient: Client,
+    FileCrypto: FileEncryptionService,
 > {
     metadatas: PhantomData<FileMetadataDb>,
+    files: PhantomData<FileDb>,
     accounts: PhantomData<AccountDb>,
-    client: PhantomData<C>,
+    client: PhantomData<ApiClient>,
+    file_crypto: PhantomData<FileCrypto>,
 }
 
-impl<FileMetadataDb: FileMetadataRepo, AccountDb: AccountRepo, ApiClient: Client>
-    FileMetadataService for FileMetadataServiceImpl<FileMetadataDb, AccountDb, ApiClient>
+impl<
+        FileMetadataDb: FileMetadataRepo,
+        FileDb: FileRepo,
+        AccountDb: AccountRepo,
+        ApiClient: Client,
+        FileCrypto: FileEncryptionService,
+    > FileMetadataService
+    for FileMetadataServiceImpl<FileMetadataDb, FileDb, AccountDb, ApiClient, FileCrypto>
 {
-    fn update(db: &Db) -> Result<Vec<FileMetadata>, Error> {
+    fn sync(db: &Db) -> Result<Vec<FileMetadata>, Error> {
+        // Load user's account
         let account = AccountDb::get_account(&db)?;
-
+        // Get the last synced file
         let max_updated = match FileMetadataDb::last_updated(db) {
             Ok(max) => max,
             Err(_) => 0,
         };
-
-        let updates = ApiClient::get_updates(
-            API_LOC.to_string(),
-            &GetUpdatesRequest {
-                username: account.username.to_string(),
-                auth: "".to_string(),
-                since_version: max_updated as u64,
-            },
-        )
-        .map(|metadatas| {
-            metadatas
-                .into_iter()
-                .map(|t| FileMetadata {
-                    id: t.file_id,
-                    name: t.file_name,
-                    path: t.file_path,
-                    updated_at: t.file_metadata_version as i64,
-                    // TODO: Fix this so status is tracked accurately
-                    status: "Remote".to_string(),
-                })
-                .collect::<Vec<FileMetadata>>()
+        // Get remote updates from the last synced file onwards
+        info(format!("Getting updates past {}", max_updated));
+        let updates_remote = ApiClient::get_updates(&GetUpdatesRequest {
+            username: account.username.to_string(),
+            auth: "JUNK_AUTH".to_string(),
+            since_version: max_updated,
         })?;
+        debug(format!("Remote Updates {:?}", updates_remote));
 
-        updates
-            .into_iter()
-            .for_each(|meta| match FileMetadataDb::insert(&db, &meta) {
-                Ok(_) => {}
-                Err(err) => {
-                    error(format!("Insert Error {:?}", err));
+        // Create all the "new" files
+        let mut all_files = FileMetadataDb::get_all(&db)?;
```

Comment:
I'm sure this could be done without `mut`. Keeping track of all the changes to `all_files` could be hard to reason about. Surely there's some way to do something like:
```
all_files
.filter
.for_each
...
```